text/0000-inline-asm.md

```diff
@@ -0,0 +1,1055 @@
+  - The stack pointer must be restored to its original value before leaving the asm block.
+- If the `noreturn` option is set then behavior is undefined if execution falls through to the end of the asm block.
+- If the `pure` option is set then behavior is undefined if the `asm` has side-effects other than its direct outputs. Behavior is also undefined if two executions of the `asm` code with the same inputs result in different outputs.
+  - When used with the `nomem` option, "inputs" are just the direct inputs of the `asm!`.
+  - When used with the `readonly` option, "inputs" comprise the direct inputs of the `asm!` and any memory that the `asm!` block is allowed to read.
+- These flags registers must be restored upon exiting the asm block if the `preserves_flags` option is set:
+  - x86
+    - Status flags in `EFLAGS` (CF, PF, AF, ZF, SF, OF).
+    - Direction flag in `EFLAGS` (DF).
+    - Floating-point status word (all).
+    - Floating-point exception flags in `MXCSR` (PE, UE, OE, ZE, DE, IE).
+  - ARM
+    - Condition flags in `CPSR` (N, Z, C, V)
+    - Saturation flag in `CPSR` (Q)
+    - Greater than or equal flags in `CPSR` (GE).
+    - Condition flags in `FPSCR` (N, Z, C, V)
+    - Saturation flag in `FPSCR` (QC)
+    - Floating-point exception flags in `FPSCR` (IDC, IXC, UFC, OFC, DZC, IOC).
+  - AArch64
+    - Condition flags (`NZCV` register).
+    - Floating-point status (`FPSR` register).
+  - RISC-V
+    - Floating-point exception flags in `fcsr` (`fflags`).
+- The requirement of restoring the stack pointer and non-output registers to their original value only applies when exiting an `asm!` block.
+  - This means that `asm!` blocks that never return (even if not marked `noreturn`) don't need to preserve these registers.
+  - When returning to a different `asm!` block than you entered (e.g. for context switching), these registers must contain the value they had upon entering the `asm!` block that you are *exiting*.
+    - You cannot exit an `asm!` block that has not been entered. Neither can you exit an `asm!` block that has already been exited.
+    - You are responsible for switching any target-specific state (e.g. thread-local storage, stack bounds).
+    - The set of memory locations that you may access is the intersection of those allowed by the `asm!` blocks you entered and exited.
+
+> Note: As a general rule, these are the flags which are *not* preserved when performing a function call.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Unfamiliarity
+
+This RFC proposes a completely new inline assembly format.
+It is not possible to just copy examples of GCC-style inline assembly and re-use them.
+There is however a fairly trivial mapping between the GCC-style and this format that could be documented to alleviate this.
+
+Additionally, this RFC proposes using the Intel asm syntax on x86 instead of the AT&T syntax. We believe this syntax will be more familiar to most users, but may be surprising for users used to GCC-style asm.
```

Comment:
The intel syntax is the one sanctioned in the intel x86 documentation, and most of the inline asm I've seen in Rust sets the `intel` flag instead of keeping the AT&T default. MSVC does support assembly through the MASM tool (not inline asm though), and MASM defaults to intel syntax. 32-bit x86 in MSVC also defaults to intel syntax.

Reply:
I don't think it matters what non-inline assemblers do. The RFC is trying to make a familiarity argument, but I don't think it holds up. Familiarity can't be supported by anecdotal evidence. Unless we do a thorough analysis of inline assembly across open source projects in different languages, the only thing we have to go on is what other compilers do. For 64-bit x86, there's a clear default.

Reply:
Familiarity isn't segmented by inline vs stand-alone assemblers, though. "Intel syntax" and "AT&T syntax" are basically the same in both scenarios so any familiarity gained in one scenario applies to the other.